docs/adr/003-knooppunt-solution-architecture.md

```diff
@@ -0,0 +1,132 @@
+
+Note that Oasis specifies more roles (PIP, PAP), but those are not relevant for this ADR.
+
+### Design Goals
+We want a solution that is easy to integrate in varying (existing) environments, without compromising on security and simplicity. 
+
+- **Simplicity**: Easy to deploy and manage, with minimal configuration required. Prevent vendor-specific integration.
+- **Flexibility**: Can be adapted to different deployment environments and use cases.
+- **Modularity**: Components (e.g. Nuts node, Open Policy Agent) can be enabled/disabled as needed.
+- **Security**: Minimize attack surfaces.
+- **Pluggability**: Should be as easy as possible to integrate.
+
+### Use cases
+
+Deciding on the architecture depends on the use cases we want to support. The main use cases are:
+
+- Handling data exchange, initiated by a remote party (inbound)
+  - This involves receiving requests, authenticating, authorizing, and responding.
+  - It could also involve masking or filtering data based on consent or authorization rules, before it's returned to the requester.
+- Initiating data exchange with a system from a remote party (outbound)
+    - This involves looking up parties, endpoints, authentication
+    - Outbound requests are out of scope for this ADR.
+
+## Considered Options
+This section describes considered architecture options.
+
+In any of the options, there's a proxy (e.g. NGINX, HAProxy, Traefik) in front of the Knooppunt and/or EHR FHIR API.
+This is a typical reverse proxy, handling TLS termination, routing, load balancing, caching, etc.
+
+### Knooppunt as both PEP and PDP
+The Knooppunt sits on the edge of the vendor's network, handling all data exchanges. It acts as:
+- OAuth2 Authorization Server
+- Policy Decision Point
+- Policy Enforcement Point
+
+```text
+┌─────────────────┐       ┌─────────┐       ┌────────────┐       ┌──────────────┐
+│                 │       │         │       │            │       │              │
+│ External System ├──────►│  Proxy  │──────►│ Knooppunt  ├──────►│ EHR FHIR API │
+│                 │       │         │       │ [PEP/PDP]  │       │              │
+└─────────────────┘       └─────────┘       └────────────┘       └──────────────┘
+```
+
+Data exchanges are routed through the Knooppunt, theoretically offloading all "complicated" concerns from the vendor.
+
+- Advantages:
+  - Simpler deployment, since it only requires the Knooppunt to be deployed.
+  - Easier to configure, since all configuration is centralized in the Knooppunt.
+  - Easier to reason about, since all data exchanges go through a single component.
+- Disadvantages:
+  - Large attack surface on the Knooppunt, since it handles all data exchanges.
+    Makes it harder to use security measures vendors already have in place, especially for resource transformation and filtering.
+  - Might not actually make things easier for vendors, if they want to implement requirements not supported by the Knooppunt (e.g. auditing, data minimization)
```

Comment:
```suggestion
- Advantages:
  - Simple deployment, since it only requires the Knooppunt to be deployed.
  - Easy to configure, since all configuration is centralized in the Knooppunt.
  - Easy to reason about, since all data exchanges go through a single component.
- Disadvantages:
  - Large attack surface on the Knooppunt, since it handles all inbound requests.
    Makes it harder to use security measures vendors already have in place, especially for resource transformation and filtering.
  - Less flexible when the vendor needs additional concerns not supported by the Knooppunt (e.g. auditing, data minimization)
```